utils/writeable/src/fallible.rs

```diff
@@ -0,0 +1,660 @@
+            Ok(Ok(())) => Ok(Cow::Owned(output)),
+            Ok(Err(e)) => Err(e),
+            Err(core::fmt::Error) => unreachable!("writing to string is infallible"),
+        }
+    }
+
+    /// Compares the contents of this `Writeable` to the given bytes
+    /// without allocating a String to hold the `Writeable` contents.
+    ///
+    /// This is a low-level function for implementors. For more details, see [`FallibleWriteable`].
+    fn fallible_write_cmp_bytes<E, L: Writeable, F: FnMut(Self::Error) -> Result<L, E>>(
+        &self,
+        other: &[u8],
+        handler: F,
+    ) -> Result<core::cmp::Ordering, E> {
+        let mut wc = cmp::WriteComparator::new(other);
+        match self.fallible_write_to(&mut wc, handler) {
+            Ok(Ok(())) => (),
+            Ok(Err(e)) => return Err(e),
+            Err(core::fmt::Error) => unreachable!("writing to string is infallible"),
+        }
+        Ok(wc.finish().reverse())
+    }
+}
+
+/// A writeable type with its own configurable error handling.
+///
+/// This trait offers four ways to consume errors:
+///
+/// - [`TryWriteable::checked`]
+/// - [`TryWriteable::lossy`]
+/// - [`TryWriteable::panicky`]
+/// - [`TryWriteable::gigo`]
+///
+/// This trait is blanked-implemented on all types implementing [`FallibleWriteable`],
+/// which is the lower-level trait that custom writeables should implement.
+pub trait TryWriteable: FallibleWriteable {
+    /// Handle errors with functions returning [`Result`].
+    ///
+    /// # Examples
+    ///
+    /// Successful behavior on `Option<T>`, which implements [`TryWriteable`]:
+    ///
+    /// ```
+    /// use writeable::TryWriteable;
+    ///
+    /// let try_writeable = Some("example");
+    /// assert_eq!(
+    ///     try_writeable.checked().try_write_to_string().unwrap(),
+    ///     "example"
+    /// );
+    /// ```
+    ///
+    /// Failure behavior:
+    ///
+    /// ```
+    /// use writeable::TryWriteable;
+    ///
+    /// let try_writeable = None::<&str>;
+    /// assert!(matches!(
+    ///     try_writeable.checked().try_write_to_string(),
+    ///     Err(_),
+    /// ));
+    /// ```
+    #[inline]
+    fn checked(&self) -> CheckedWriteable<&Self> {
+        CheckedWriteable(self)
+    }
+    /// Handle errors by writing a replacement to the output string.
+    ///
+    /// # Examples
+    ///
+    /// Successful behavior on `Option<T>`, which implements [`TryWriteable`]:
+    ///
+    /// ```
+    /// use writeable::TryWriteable;
+    ///
+    /// let try_writeable = Some("example");
+    /// writeable::assert_writeable_eq!(try_writeable.lossy(), "example");
+    /// ```
+    ///
+    /// Failure behavior:
+    ///
+    /// ```
+    /// use writeable::TryWriteable;
+    ///
+    /// let try_writeable = None::<&str>;
+    /// writeable::assert_writeable_eq!(try_writeable.lossy(), "�");
+    /// ```
+    #[inline]
+    fn lossy(&self) -> LossyWriteable<&Self> {
```

Comment:
`LossyWriteable` needs `Self::Error: Writeable` to work, but the bound is not here. This sounds like it will lead to very annoying errors.

I do find it a bit odd that lossy allows custom errors to be written, it's not really lossy then because no information is lost. If it unconditionally wrote U+FFFD, it'd be lossy.